### PR TITLE
Manual silu quant fusion

### PR DESCRIPTION
**Pull request by ElizaWszola**
Adding more quant types to SiLU+quant activation.

Made with the assistance of Claude, all generated code has been human-reviewed.

#### Unit test
```
pytest tests/kernels/test_fused_quant_activation.py
```

#### E2E test example (fp8, static) - manually expect if outputs are sane
```
from vllm import LLM, SamplingParams

# Sample prompts.
prompts = [
    "Hello, my name is",
    "The president of the United States is",
    "The capital of France is",
    "The future of AI is",
]
# Create a sampling params object.
sampling_params = SamplingParams(temperature=0.8, top_p=0.95, min_tokens=200, max_tokens=200)

llm = LLM(model='neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8')
outputs = llm.generate(prompts, sampling_params)
# Print the outputs.
for output in outputs:
    prompt = output.prompt
    generated_text = output.outputs[0].text
    print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")

```
#### Eval example (fp8, static)
```
lm_eval --model vllm \
    --model_args pretrained=neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8 \
    --tasks hellaswag,arc_easy \
    --batch_size auto
```

### Eval results:

(gsm8k with `neuralmagic/Meta-Llama-3.1-8B-Instruct-FP8`)

main:
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7710|±  |0.0116|
|     |       |strict-match    |     5|exact_match|↑  |0.6922|±  |0.0127|
```

this PR + https://github.com/vllm-project/vllm/pull/46864
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7582|±  |0.0118|
|     |       |strict-match    |     5|exact_match|↑  |0.6816|±  |0.0128|
```

### CODE REVIEW

**Comment by ElizaWszola** [comment]
I had initially set a wrong base commit for this PR's branch which resulted in lots of incorrect labels, would be grateful if someone could manually delete them

File: tests/kernels/test_fused_quant_activation.py [hunk 1]
```diff
@@ -69,3 +80,158 @@ def test_silu_and_mul(
...
+
+    def __init__(self):
+        super().__init__()
+        self.input_quant_key = kFp8Dynamic64Sym
+
+
+class MockLinearNoQuant(torch.nn.Module):
+    """Mock linear layer with no input_quant_key (no fusion)."""
+
+    pass
```
Comment (mgoin) [inline tests/kernels/test_fused_quant_activation.py:118]: I think we need to update the existing fusion tests, since if we are replacing these ops with manual fusions now, the torch.compile based replacements will stop happening. So we should make something more like tests/compile/passes/test_silu_mul_quant_fusion.py, replacing that with the manual equivalent

File: vllm/model_executor/layers/fusion/fused_act_quant.py [hunk 2]
```diff
@@ -49,12 +54,106 @@ def _silu_and_mul_fp8_static(
...
 _FUSED_ACT_QUANT: dict[tuple[type, QuantKey], Callable] = {
     (SiluAndMul, kFp8StaticTensorSym): _silu_and_mul_fp8_static,
 }
 
+# Add CUDA-specific entries for dynamic block quantization
+if current_platform.is_cuda_alike():
+    _FUSED_ACT_QUANT.update(
+        {
+            (SiluAndMul, kFp8Dynamic128Sym): _silu_and_mul_fp8_dynamic_128,
+            (SiluAndMul, kFp8Dynamic64Sym): _silu_and_mul_fp8_dynamic_64,
```
Comment (mgoin) [inline vllm/model_executor/layers/fusion/fused_act_quant.py]: Let's remove the block-64 case since we don't have a model we know that uses this

File: vllm/model_executor/layers/fusion/fused_act_quant.py [hunk 3]
```diff
@@ -49,12 +54,106 @@ def _silu_and_mul_fp8_static(
...
+    # Block scale output shape: swizzled layout for tensor cores
+    # Each group of 16 elements shares one FP8 scale
+    num_k_tiles = (d + 63) // 64
+    block_scale = torch.empty(
+        (num_tokens, num_k_tiles * 4), dtype=FP8_DTYPE, device=x.device
+    )
+
+    input_global_scale = getattr(linear, "input_global_scale", None)
+    if input_global_scale is None:
+        input_global_scale = torch.tensor([1.0], dtype=torch.float32, device=x.device)
```
Comment (mgoin) [inline vllm/model_executor/layers/fusion/fused_act_quant.py]: I think we should assert input_global_scale is not None, not silently make it unscaled

File: tests/kernels/test_fused_quant_activation.py (hunk 1 above)
Comment (ElizaWszola) [reply tests/kernels/test_fused_quant_activation.py:118]: I changed the tests a bit, I tried to give them similar coverage to the fusion pass ones: the new tests are in `tests/compile/passes/test_silu_mul_quant_manual_fusion.py`, can you have a look?

**Comment by ElizaWszola** [comment]
Merged the changes into https://github.com/vllm-project/vllm/pull/51415, will work on it from now
